src/app/clusters/operational-state-server/operational-state-delegate.h

```diff
@@ -0,0 +1,393 @@
+constexpr size_t kOperationalPhaseNameMaxSize    = 64u;
+
+/**
+ * A class wrap the operational state of operational state cluster
+ */
+struct GenericOperationalState : public chip::app::Clusters::detail::Structs::OperationalStateStruct::Type
+{
+    GenericOperationalState(uint8_t state, const char * label = nullptr, size_t labelLen = 0) { set(state, label, labelLen); }
+
+    GenericOperationalState(const GenericOperationalState & op)
+    {
+        set(op.operationalStateID, op.operationalStateLabel.HasValue() ? op.OperationalStateLabel : nullptr,
+            op.operationalStateLabel.HasValue() ? sizeof(op.OperationalStateLabel) : 0);
+    }
+
+    GenericOperationalState & operator=(const GenericOperationalState & op)
+    {
+        set(op.operationalStateID, op.operationalStateLabel.HasValue() ? op.OperationalStateLabel : nullptr,
+            op.operationalStateLabel.HasValue() ? sizeof(op.OperationalStateLabel) : 0);
+
+        return *this;
+    }
+
+    void set(uint8_t state, const char * label = nullptr, size_t labelLen = 0)
+    {
+        operationalStateID = state;
+        if (label == nullptr)
+        {
+            operationalStateLabel = Optional<chip::CharSpan>::Missing();
+        }
+        else
+        {
+            memset(OperationalStateLabel, 0, sizeof(OperationalStateLabel));
+            if (labelLen > kOperationalStateLabelMaxSize)
+            {
+                memcpy(OperationalStateLabel, label, kOperationalStateLabelMaxSize);
+            }
+            else
+            {
+                memcpy(OperationalStateLabel, label, labelLen);
+            }
+
+            operationalStateLabel.SetValue(chip::CharSpan(OperationalStateLabel, sizeof(OperationalStateLabel)));
+        }
+    }
+    uint8_t getStateID() const { return operationalStateID; }
+    char OperationalStateLabel[kOperationalStateLabelMaxSize];
+};
+
+/**
+ * A class hold the operational state list of operational state cluster
+ */
+struct GenericOperationalStateList : public GenericOperationalState
+{
+    GenericOperationalStateList(uint8_t state, const char * label = nullptr, size_t labelLen = 0) :
+        GenericOperationalState(state, label, labelLen)
+    {}
+    GenericOperationalStateList * next = nullptr;
+};
+
+/**
+ * A class wrap the operational error of operational state cluster
+ */
+struct GenericOperationalError : public chip::app::Clusters::detail::Structs::ErrorStateStruct::Type
+{
+    GenericOperationalError(uint8_t state, const char * label = nullptr, size_t labelLen = 0, const char * details = nullptr,
+                            size_t detailsLen = 0)
+    {
+        set(state, label, labelLen, details, detailsLen);
+    }
+
+    GenericOperationalError(const GenericOperationalError & error)
+    {
+        set(error.errorStateID, error.errorStateLabel.HasValue() ? error.ErrorStateLabel : nullptr,
+            error.errorStateLabel.HasValue() ? sizeof(error.ErrorStateLabel) : 0,
```

Comment:
Same comments as for operational state apply, including all the lengths being wrong.

I am going to stop reviewing now.  This needs tests, which actually pass, because trying to catch all these issues during review is just going to make things slow and be error-prone.